Make tests independent

kamyar opened the pull request:
fixes #170

Tests functionally should be testing the exact same thing as before, main change is side effects of previous tests and setup/teardown methods were moved to fixtures in new conftest.py file, there are some extra changes that came because of this(pylint mainly not liking method tests anymore because they can also be simple functions etc.)

Would appreciate any comments/suggestions. :)


Changes:
- move index creation, cleanup of test run, reading small movies and adding them to index to fixtures
- add an extra fixture to cleanup data after each tests (just to be safe for tests that will come later on, kind of meta test I guess?)
- remove setup and teardown methods, use the new fixtures where ever they were needed
- now that having tests do not need a class, pylint was getting sad, converted many tests to be simple pytest function tests rather than method test.
    - this made pylint still sad because function names for tests are > 30 chars, updated that to 70 chars
    - constants that were class attributes are causing pylint errors, made them SCREAMING_SNAKE_CASE 
- created tests/common.py for common constant values(e.g. index configs, sample index names etc.)
- made meili client a fixture scoped to test session, one client creation should be enough and test can easily access it by using the `client` fixture.

TODO:
- [x] Add detailed PR description about the changes
- [x] use `@pytest.mark.usefixtures` for unused fixtures for happier linter (https://github.com/PyCQA/pylint/issues/1057#issuecomment-237155725)
- [x] try moving tests/__init__.py code to tests/common.py
- [x] Check leftover tests? (even if they are passing without any change try to unify the structure of tests)

curquiza commented on meilisearch/tests/conftest.py:
```diff
@@ -0,0 +1,71 @@
+@fixture(scope="function")
+def custom_indexed_maker(client, small_movies):
+    def index_maker(index_suffix):
+        index = client.create_index(uid='indexUID-' + index_suffix)
```
Now, MeiliSearch automatically creates the index if it does not exist during the document addition! The `index()` method only returns an instance of `Index` and the real index will be created during the line execution below.
You started your PR before we merged this change so you might have not noticed it! See the new getting started -> https://github.com/meilisearch/meilisearch-python#-getting-started

curquiza replied:
(Even if I ask to remove `custom_indexed_maker`, I think this remark is interesting to implement `index_with_documents` fixture)

kamyar replied:
just tried this but got the following error:
```
meilisearch.errors.MeiliSearchApiError: MeiliSearchApiError. Error code: index_not_found. Error message: Index indexUID not found. Error documentation: https://docs.meilisearch.com/errors#index_not_found
```

(also pulled docker but it seems I already had the up-to-date version)

Any suggestions what I need to update to make it work?

curquiza replied:
Weird! Did you use the v0.16.0 version of MeiliSearch?

curquiza commented on meilisearch/tests/conftest.py:
```diff
@@ -0,0 +1,71 @@
+def custom_indexed_maker(client, small_movies):
+    def index_maker(index_suffix):
+        index = client.create_index(uid='indexUID-' + index_suffix)
+        response = index.add_documents(small_movies, primary_key='id')
```
```suggestion
        response = index.add_documents(small_movies)
```

curquiza replied:
No need to use the `primarey_key` options, MeiliSearch will infer it since it's a name with `id` inside 😉
https://docs.meilisearch.com/guides/main_concepts/documents.html#meilisearch-infers-your-primary-key

curquiza replied:
(Even if I ask to remove `custom_indexed_maker`, I think this remark is interesting to implement `index_with_documents` fixture)

curquiza commented on meilisearch/tests/conftest.py:
```diff
@@ -0,0 +1,71 @@
+        index.wait_for_pending_update(response['updateId'])
+        yield
+        # cleanup
+        index.delete()
```
Same as for `sample_indexes`, no need to clean it up from my POV, of maybe I do not understand well something.

kamyar replied:
Sounds good, this is also redundant. Will remove this also 👍

curquiza commented on meilisearch/tests/index/test_index.py:
```diff
@@ -1,160 +1,155 @@
-        assert len(self.client.get_indexes()) == 0
+from meilisearch.tests import common
+
+
+
```
```suggestion
```

curquiza reviewed (changes requested):
Thank you so much @kamyar of this huuuuuuuge PR 🙏 
Make these tests independent was one of our main concerns and you saved us so much time!!

I have some requested changes, but keep in mind your work was really clear to me: the comments you added and your detailed PR description made this PR really pleasant to review. If all the contributor PRs were like you, it would be the best opensource world! 🚀 

About the code:
- [ ] I would add an`empty_index` and an `index_with_documents` fixtures. `sample_indexes` is a useful fixture, for example when testing the get of all the indexes. But it is not really convenient when testing the search, the settings or the documents manipulation, I mean usecases when you only need ONE index, because you have to write `sample_indexes[0].method_name` instead of `index_with_documents.method_name`.
- [ ] Related to the point above: I would remove the `indexed_small_movies` fixture that is only used with the first index of the array (`indexed_small_movies[0]`). `index_with_documents` would be enough to perform these tests.
- [ ] I would remove the `custom_indexed_maker` that is only used for the dumps but that is not necessary in this case: the `empty_index` or `index_with_documents` fixture are enough to test it.
- [ ] Rename `sample_indexes` by `indexes_sample`
- [ ] I also add some remarks directly in the code review

About styling (sorry to bother you with these):
- [ ] You used tabs in the conftest.py file. We used spaces in this project 😉.
- [ ] We use single quotes `'` instead of double quotes `"` for literal strings. For example: `@fixture(scope="session")` -> `@fixture(scope='session')`

About git:
You make your branch pointing to `staging`, but we will directly merge it into `master`. See our contribution guidelines -> https://github.com/meilisearch/meilisearch-python/blob/master/CONTRIBUTING.md#how-to-contribute
**I've already made the change, so you don't have to change anything about this**, I just inform you 🙂
FYI, the `staging` branch is used by one of our tools (bors), and I haven't documented it in this repository yet, sorry, so you couldn't guess, that's my fault 😅

Thanks again for this PR, I learned a lot about pytest and fixtures thanks to you 👍

kamyar commented on .pylintrc:
```diff
@@ -177,7 +177,7 @@ name-group=
 
 # Regular expression matching correct function names
-function-rgx=[a-z_][a-z0-9_]{2,30}$
+function-rgx=[a-z_][a-z0-9_]{2,70}$
```
Makes sense, will try to make pylint treat `tests/*` differently maybe even or will add the disable comment if that is not possible. 👍

kamyar commented on meilisearch/tests/conftest.py:
```diff
@@ -0,0 +1,71 @@
+    # yield the indexes to the test so it can use it
+    yield indexes
+    # test finished, let's cleanup
+    for index in indexes:
+        try:
+            index.delete()
+        except meilisearch.errors.MeiliSearchApiError:
+            # test deleted index explicitly
+            pass
```
I think I added it before I fixed some of the tests, you are right, it already clears everything already, this is redundant, will remove 👍

kamyar commented:
Thank you @curquiza , great to hear PR was pleasant 🎉 
I am very happy to hear that you learnt from this PR. 😍 

I tried to address all of your comment, please let me know if I missed something/over did it. :) 

Some notes:
- [ ] tabs seemed to be in common.py rather than conftest.py, fixed that, may I recommend adding something like [pre-commit](https://pre-commit.com/)? have been very useful at work since we started using it.
- [ ] removed redundant index removals, you are right, we already remove them with autouse, so that should be enough.
- [ ] I might not fully understood `empty_index` and `index_with_documents` and other fixture related naming/structure changes, but tried to achieve what I thought you wanted, hopefully it matches. please lett me know if I missed something here.
- [ ] use of single quote instead of double 
- [ ] reverted the function name pylint change, added disable comment

curquiza commented on meilisearch/tests/conftest.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@fixture(scope='function')
+def empty_index(client):
+    def index_maker(index_name=common.INDEX_UID):
+        return client.create_index(uid=index_name)
+    return index_maker
```
Sorry not to have asked before, but with my colleague @eskombro we were asking what is the purpose of doing a function in another function? 🙂

kamyar replied:
it is creating higher order function, you predefine some things and pass some other things later on.

in pytest-context:
fixtures already get called before a test runs, and test gets the result.
But what if we want a fixture that does something slightly differently depending on a param our tests passes?
We can define a fixture(empty_index), that returns a function to make indexes.

In this specific case, `index_name` by default has value, but maybe a function uses this multiple times to create many indexes with slightly different names: `index_maker("index-1") then index_maker("index-2")`

It can be avoided of course, test can import a simple helper function, that function imports `(meili) client` and depending on parameters that the test has given it creates an index.
But this approach can be applied for any fixture, every test can call helper functions to achieve stuff instead of fixture, just wanted to stick to the fixture way of doing it.

function-in-function here can be I think thought of very similar(even same) thing to python decorators do or javascript HOCs.

Please let me know if I was ambiguous or missed the point :D

curquiza replied:
Got it!! Thanks!

curquiza commented on meilisearch/tests/settings/test_settings.py:
```diff
@@ -1,72 +1,61 @@
+
+def test_reset_settings(empty_index):
+    """Tests resetting default settings"""
+    index = empty_index()
+    response = index.reset_settings()
```
I should have asked it before, so sorry about this!
I just realized this test is not well done anymore since the settings are not updated before resetting them!

curquiza replied:
```suggestion
    index = empty_index()
    # Update settings first
    response = index.update_settings(NEW_SETTINGS)
    update = index.wait_for_pending_update(response['updateId'])
    assert update['status'] == 'processed'
    # Check the settings have been correctly updated
    response = index.get_settings()
    for rule in NEW_SETTINGS['rankingRules']:
        assert rule in response['rankingRules']
    assert response['distinctAttribute'] is None
    for attribute in NEW_SETTINGS['searchableAttributes']:
        assert attribute in response['searchableAttributes']
    assert response['displayedAttributes'] == ['*']
    assert response['stopWords'] == []
    assert response['synonyms'] == {}
    # Check the reset of the settings
    response = index.reset_settings()
```

curquiza commented on meilisearch/tests/settings/test_settings_attributes_for_faceting_meilisearch.py:
```diff
@@ -1,48 +1,30 @@
+
+def test_reset_attributes_for_faceting(empty_index):
+    """Tests the reset of attributes for faceting to default values (in dataset)"""
+    index = empty_index()
+    response = index.reset_attributes_for_faceting()
```
Same than above

curquiza commented on meilisearch/tests/settings/test_settings_displayed_attributes_meilisearch.py:
```diff
@@ -1,50 +1,35 @@
+
+def test_reset_displayed_attributes(empty_index):
+    """Tests the reset of displayedAttributes to default values (in dataset)"""
+    index = empty_index()
+    response = index.reset_displayed_attributes()
```
Same than above

curquiza commented on meilisearch/tests/settings/test_settings_distinct_attribute_meilisearch.py:
```diff
@@ -1,42 +1,31 @@
+
+def test_reset_distinct_attribute(empty_index):
+    """Tests resetting distinct attribute"""
+    index = empty_index()
+    response = index.reset_distinct_attribute()
```
Same than above

curquiza commented on meilisearch/tests/settings/test_settings_ranking_rules_meilisearch.py:
```diff
@@ -1,52 +1,40 @@
+
+def test_reset_ranking_rules(empty_index):
+    """Tests resetting the ranking rules"""
+    index = empty_index()
+    response = index.reset_ranking_rules()
```
Same than above

curquiza commented on meilisearch/tests/settings/test_settings_searchable_attributes_meilisearch.py:
```diff
@@ -1,54 +1,37 @@
+
+def test_reset_searchable_attributes(empty_index):
+    """Tests reseting searchable attributes"""
+    index = empty_index()
+    response = index.reset_searchable_attributes()
```
Same than above

curquiza commented on meilisearch/tests/settings/test_settings_stop_words_meilisearch.py:
```diff
@@ -1,44 +1,32 @@
+
+def test_reset_stop_words(empty_index):
+    """Tests call to reset stop words"""
+    index = empty_index()
+    response = index.reset_stop_words()
```
Same than above

curquiza commented on meilisearch/tests/settings/test_settings_synonyms_meilisearch.py:
```diff
@@ -1,46 +1,34 @@
+
+def test_reset_synonyms(empty_index):
+    """Tests resetting synonyms"""
+    index = empty_index()
+    response = index.reset_synonyms()
```
Same than above

curquiza reviewed (changes requested):
Thanks for your changes @kamyar! 😍 

> I might not fully understood empty_index and index_with_documents and other fixture related naming/structure changes, but tried to achieve what I thought you wanted, hopefully it matches. please lett me know if I missed something here.

That was exactly what I meant! 👍 

> tabs seemed to be in common.py rather than conftest.py, fixed that, may I recommend adding something like pre-commit? have been very useful at work since we started using it.

Sorry, I miswrote the name of the file!
And about the pre-commit, that's a good tool! However, we have already thought about this, and we are afraid this discourages people to submit their PR if there is a hook that avoids them to push and commit. But, on the other hand, we know this could be more convenient for contributors like you. Anyway, be sure this is one of our concern, and I keep your suggestion in mind 😉 

About the PR: I only ask a question and another change, but that's my fault I should have asked this change before! The change is about "reset settings" tests. Since the tests are now independent, the reset of the setting is not well tested anymore: we need to update and then, check the settings are well updated before resetting. Otherwise, we could not see the reset was well applied.
I made a concrete example for the first one so that you can see what I mean. But this involves changes for the other settings files... **If it's too much work for you or you don't have the time, let me know and I will do a PR to your branch with the modifications!** 😄 

Thanks again for your patience and your involvement! That's a pleasure to work with people like you! 🚀

kamyar commented:
@curquiza thank you for the comment and review.
I tried to answer the question, regarding the setting test fix I applied the suggested edit you made, but of course feel free to push to the branch, I don't if you do, especially as I have less business-logic context than you obviously. 😅 
It will be actually better if you made the other setting changes just in case I miss something. wdyt?
I am happy to do my best if you are already busy.

Regarding pre-commit, you can set it up such that it automatically fixes the mismatches and shows errors on commit, I think it would not be too much headache for anyone even potential new contributors. :) 

It can be simple also, just `tabs -> spaces` and `removal of trailing whitespaces` etc.

curquiza commented:
I'm going to merge this PR as it is then, and I'm going to open a PR right now to update the tests :smile:

Thanks for everything @kamyar! This PR is a real improvement for our team, we really thank you!

curquiza reviewed (approved):
bors merge

eskombro commented:
Thanks for this contribution @kamyar 🎉

kamyar commented:
Thank you all @curquiza & @eskombro for the support, was a very delightful experience from my side also. 🎉 
Especially your humbleness and support throughout the PR, hope I will be able to contribute more, looking forward to using meili search in a project also, looks like a great project!

curquiza commented:
I'm really glad to read contributing to a MeiliSearch tool was a great experience for you! It was also really pleasant for us to work with you :wink:

Don't hesitate to join us on \[our Slack\](https://slack.meilisearch.com/), if you're not already there! I'm Clementine.
